### ISSUE TITLE
Gas costs post-memtrie

### ISSUE THREAD

**Issue by Ekleog-NEAR** [near/nearcore#11925]
With memtrie, we made our storage reads much faster. It already enabled stateless validation.

Now we want to lower at least compute costs, in order to also get single-shard performance improvements from that work.

In order to do that, we must retake ownership of parameters estimator, run it with the newest changes, and then consider what costs changes we should do.

**Comment by Ekleog-NEAR** [comment]
### Data out of the runs (without memtrie enabled yet)

Skill unlocked: using chatgpt to turn unstructured data into structured data:

Iteration | StorageHasKeyBase (Gas) | StorageHasKeyByte (Gas) | StorageReadBase (Gas) | StorageReadKeyByte (Gas) | StorageReadValueByte (Gas) | StorageWriteBase (Gas) | StorageWriteKeyByte (Gas) | StorageWriteValueByte (Gas) | StorageWriteEvictedByte (Gas) | StorageRemoveBase (Gas) | StorageRemoveKeyByte (Gas) | StorageRemoveRetValueByte (Gas) | TouchingTrieNode (Gas) | ReadCachedTrieNode (Gas) | Computation Time (Seconds)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 12,900,400,629 | 32,974,229 | 13,176,595,629 | 32,950,083 | 1,996,159 | 5,894,729,584 | 6,054,362 | 2,372,167 | 3,502,758 | 13,325,130,629 | 12,766,994 | 1,660,624 | 8,087,650,976 | 838,062,500 | 879.35
2 | 15,193,494,159 | 35,605,815 | 13,538,829,159 | 33,160,425 | 2,006,289 | 6,010,015,762 | 6,186,676 | 2,401,317 | 3,662,605 | 13,380,131,159 | 12,818,790 | 1,669,679 | 6,168,517,526 | 938,437,500 | 867.08
3 | 13,065,329,313 | 32,983,184 | 13,752,679,313 | 33,064,852 | 2,080,969 | 5,988,201,695 | 6,486,408 | 2,391,904 | 3,559,623 | 13,027,891,147 | 12,733,717 | 1,692,055 | 8,995,546,152 | 897,291,666 | 960.9
4 | 14,058,590,550 | 33,011,596 | 14,448,919,175 | 32,979,505 | 2,017,662 | 6,040,652,050 | 6,248,627 | 2,347,736 | 3,590,032 | 13,096,787,050 | 12,601,924 | 1,640,915 | 6,211,604,656 | 971,890,625 | 1034.3
5 | 13,483,639,212 | 32,955,855 | 14,311,978,112 | 33,267,940 | 2,027,850 | 6,025,332,562 | 6,184,661 | 2,391,077 | 3,774,544 | 14,726,878,112 | 12,753,175 | 1,704,211 | 6,588,516,675 | 1,069,487,500 | 664.6
10 | 13,578,744,617 | 32,979,644 | 13,407,996,617 | 33,113,673 | 2,069,821 | 6,324,529,522 | 6,161,299 | 2,388,617 | 3,691,684 | 13,050,943,167 | 12,708,137 | 1,690,279 | 7,055,293,690 | 1,080,375,000 | 747.11
20 | 13,534,877,379 | 32,954,422 | 13,596,178,804 | 33,125,687 | 2,107,915 | 6,532,413,647 | 6,229,705 | 2,400,606 | 3,641,082 | 13,505,353,779 | 12,778,403 | 1,705,733 | 7,165,741,437 | 990,625,000 | 1952.98
30 | 13,672,466,157 | 32,923,901 | 13,992,991,990 | 32,970,357 | 2,122,544 | 6,609,279,470 | 6,356,957 | 2,400,035 | 3,977,237 | 14,193,024,007 | 12,636,652 | 1,692,422 | 7,553,665,014 | 929,081,250 | 3679.74
40 | 13,589,294,563 | 34,641,766 | 14,583,977,776 | 32,810,104 | 2,199,171 | 6,687,153,364 | 6,298,567 | 2,415,817 | 3,950,634 | 13,956,983,851 | 12,773,246 | 1,709,642 | 13,769,501,151 | 980,657,812 | 5040.11
50 | 13,776,535,151 | 32,789,001 | 14,081,979,751 | 33,072,709 | 2,285,782 | 6,764,527,654 | 6,185,901 | 2,697,848 | 5,259,693 | 16,254,930,461 | 14,017,706 | 1,818,455 | 7,465,078,177 | 1,019,361,250 | 5999.32
75 | 14,014,411,258 | 32,971,225 | 14,365,610,898 | 33,022,814 | 2,350,673 | 7,296,011,225 | 6,315,680 | 2,538,901 | 4,342,591 | 13,523,278,652 | 12,850,151 | 1,765,050 | 7,156,997,956 | 981,690,833 | 9295.41
100 | 19,351,911,321 | 33,052,763 | 17,326,255,431 | 33,028,469 | 2,226,692 | 7,246,144,383 | 6,034,886 | 2,500,958 | 4,125,927 | 16,513,323,191 | 12,870,466 | 2,041,085 | 8,713,369,479 | 960,275,000 | 12105.58
150 | 14,480,452,840 | 33,187,697 | 18,059,659,117 | 33,280,597 | 2,325,098 | 8,026,454,400 | 6,700,411 | 2,529,798 | 4,247,221 | 19,840,170,697 | 13,153,151 | 1,979,379 | 7,543,821,926 | 1,095,268,333 | 19273.83
Current Costs | 200,000,000,000 | 30,790,845 | 200,000,000,000 | 30,952,533 | 5,611,005 | 200,000,000,000 | 70,482,867 | 31,018,539 | 32,117,307 | 200,000,000,000 | 38,220,384 | 11,531,556 | 110,000,000,000 | 2,280,000,000 |  



This gives the following graphs:

-----

![Variations around average across all runs](https://github.com/user-attachments/assets/8c1a8568-9288-4985-8433-631c9a5be8a4)


In particular, it shows that while the results are mostly stable over a different number of iterations, the remaining instability seems to not be particularly lower when running more iterations. So the sweet spot will likely be running like ten 10-iters runs and doing data magic on them to decide what's good for us.

There is a slight ramp-up when increasing the iteration number, probably due to variance being better-detected and thus pushing the numbers up. This is likely not very relevant, as we should take a safety margin of more than this number.

-----

![Estimator run vs  costs (log scale)](https://github.com/user-attachments/assets/cc742556-31af-4e5d-bf79-cdd1110a3761)

So currently, our costs are set quite a bit higher than estimator results. If I understand [Jakob's analysis](https://github.com/near/nearcore/issues/8006) correctly, this was done in order to derisk the flat storage launch: we wanted to be sure that our costs were not underestimated in the worst case.

AFAIU we're changing that policy nowadays to say that costs that just avoid missed chunks with our real-world load should be enough. Still, considering we're looking into various ways of getting state writes out of the critical path altogether, I'd suggest waiting until that happens to pursue gas costs reductions in writes: once flushing to disk is no longer a risk, we can safely ignore all the computations Jakob did in #8006 and just set values to the estimator-provided values.

-----

### Next steps

This project started around reevaluating the read costs with memtrie, so we should focus on that. But then we should likely continue with reducing the write costs too. My planned next steps are:
1. Update the estimator to enable memtrie
2. Confirm that the speedup order of magnitude is as expected
3. Pick some prospective gas costs for reads (this won't need as much care as flat storage, because memtrie performance should be much more stable and much less dependent on filesystem caches etc.)
4. Implement #11944
5. Use it to validate the prospective gas costs
6. If confirmed, perform the full protocol change and deploy the new gas costs in the next release

**Comment by Ekleog-NEAR** [comment]
Here is what the numbers of the parameter estimator out of [12adab4](https://github.com/near/nearcore/pull/11963/commits/12adab425ec90e574e1a6746d3eaeb0803b0623b) (with the log line removed) look like:

![chart (1)](https://github.com/user-attachments/assets/259b363f-a523-4456-b77b-c7a0e16a0403)

![chart (2)](https://github.com/user-attachments/assets/5a08802d-53d9-46c9-9223-1af5bccee3b8)

![chart (3)](https://github.com/user-attachments/assets/65170ca8-1dbb-4bb8-b63b-3e620a8042d2)

(The full data for the runs used for these graphs is [available here](https://docs.google.com/spreadsheets/d/108dTOXrKWDe2yBjcrpqjemljB-HEtkc9-gdf9_tLXrY/edit?usp=sharing))

So memtrie does bring in very significant benefits on estimator results! Though it also seems to make writing a bit slower at times. Thanksfully we already have a very large safety margin on writes, so we should be able to just ignore the increase in costs here. And the most significant benefit of memtrie is that it removes all disk access from the code path, meaning that the variance should be much lower, and thus that we can take a much smaller safety margin compared to the previous 15G -> 200G safety margin that we needed to account for disk limits under load.

With all that in mind, I suggest targeting a roughly 20~30% safety margin above estimator 90th percentile, because memtrie should have a pretty low variance:
- `StorageHasKeyBase`: 200,000,000,000 -> 8,000,000,000, without 4k split because checking whether the key exists should AFAIU never require falling back to disk (⚠️ assumption that needs to be confirmed by someone familiar with memtrie)
- `StorageHasKeyByte`: 30,790,845 -> 9,000,000, without 4k split (⚠️ same assumption)
- `StorageReadBase`: 200,000,000,000 -> same if the value is 4k or more; 9,000,000,000 if the value is strictly less than 4k
- `StorageReadKeyByte`: 30,952,533 -> same if the value is 4k or more; 10,000,000 otherwise
- `StorageReadValueByte`: 5,611,005 -> same if the value is 4k or more; 2,500,000 otherwise
- `StorageWrite*`: do not change
- `StorageRemove*`: do not change
- `TouchingTrieNode`: 110,000,000,000 -> 6,000,000,000, without 4k split (⚠️ same assumption)
- `ReadCachedTrieNode`: 2,280,000,000 -> 1,400,000,000, we probably do not need the 2x safety margin that we originally had, even though memtrie did not make things much faster the estimator is still around 1,070,000,000 only

My next step will be making the protocol version change that does this split. It'll be my first time adding a cost so it might take a bit longer than initially expected.

@Longarithm / @tayfunelmas can you check in particular the assumptions in three places that have the ⚠️ sign? I'll also double-check them myself while making the protocol change, but you know much better than I do about storage.

Also, what do you think about these numbers? They seem to be reasonable to me, but are they to you too?

-----

### Next steps

Considering the recent discussion, I'm removing 4 and 5 from the next steps listed above. The next steps thus are:
1. Get feedback on whether the numbers look fine, and the assumptions seem reasonable
2. Implement the protocol change that updates the current costs and adds the less-than-4k costs where applicable
3. Land the PR and rely on release mocknet testing to validate that it does not break too much.

**Comment by akhi3030** [comment]
also CC: @pugachAG

**Comment by akhi3030** [comment]
this is very good work.  I am happy to see that we are moving forward with these cost reductions.  Excited for the next steps.

**Comment by pugachAG** [comment]
@Ekleog-NEAR we also need to consider how storage operations contribute to the chunk state witness size. My intuition says that with memtrie that would mostly determine gas costs, not the latency.

**Comment by akhi3030** [comment]
Just finished a call with @ pugachAG to understand their concern above better.  

When we access storage, there are two types of costs that we need to consider.
1. the time to actually access the data and return it to the contract.  This cost has gone down for reads due to memtrie as seen above.
2. Storage accesses have to be added to the state witness.  And there is an upper limit on how bit the state witness can be.  Once, we hit this limit, we will no longer include additional transactions in the chunk.

The concern is that if we reduce gas costs based purely on 1. above, then we simply get limited by the state witness size limit.  Hence, we should consider both the limits above to decide what the new gas values should be.  I agree with this.  I don't think it makes sense to reduce gas costs a lot when we will not see significant benefit of that in practice if we keep getting limited by the state witness size.

@Ekleog-NEAR: let's discuss in more detail when you are back from your PTO.  I imagine that we should run a few more experiments to figure out how low we can make the gas costs before further reductions will not make sense.  

Another point that we discussed in the call is that the current TTN for reading cost is too low.  It was set incorrectly when we shipped flat state.  So we should actually take this opportunity to see what the correct value should be and if check if it needs to be increased even.

**Comment by akhi3030** [comment]
@Ekleog-NEAR: I had a conversation with Bowen about this work.  A question came up that it seems like we should expect memtrie to help with writes and not with reads.  Reads were already optimised due to flat storage.  

Finally, there are some other concerns with reducing the gas costs.  We should look at what the current compte costs are and instead consider reducing them instead.

**Comment by bowenwang1996** [comment]
@Ekleog-NEAR when you run the experiments, did you include https://github.com/near/nearcore/commit/0c2374993fc74b57faf2bcdf5c7c73a37e82b75a ? According to @pugachAG this significantly improves memtrie performance.

**Comment by Ekleog-NEAR** [comment]
> The concern is that if we reduce gas costs based purely on 1. above, then we simply get limited by the state witness size limit. Hence, we should consider both the limits above to decide what the new gas values should be. I agree with this. I don't think it makes sense to reduce gas costs a lot when we will not see significant benefit of that in practice if we keep getting limited by the state witness size.

I don't think I agree with this concern. Compute costs and state witness size are two orthogonal limitations, and it just happens that storage reads consume the two. But we have a lot of other operations that consume only compute costs.

So for example, imagine that we have an original workload that does 1 storage read for 1 hash computation. There, it's useless to reduce compute costs by a lot, because we'll very quickly be limited by state witness size. But then a new workload comes in, that does 1 storage read followed by 1M hash computations. If we do not reduce the compute costs, then this workload will be heavily limited by compute costs.

As we don't know what future workloads will be coming on our blockchain, and there's no value in keeping compute costs higher than they should be, I do think that we should reduce them to the lowest value we feel comfortable handling — and increasing them again is not hard anyway.

> Another point that we discussed in the call is that the current TTN for reading cost is too low. It was set incorrectly when we shipped flat state. So we should actually take this opportunity to see what the correct value should be and if check if it needs to be increased even.

@pugachAG My understanding was that TTN would only be dealing with values of less than 4k, so would literally never touch the disk trie. Is this understanding wrong?

(I also saw that TTN was set to too low a value when deploying flat storage, but if this understanding is correct then we no longer need to account for the variance there was in flat storage; and thus we can set it to almost the estimator-provided value. Said variance was the reason why TTN costs could at times have been underestimated)

-----

> @Ekleog-NEAR: I had a conversation with Bowen about this work. A question came up that it seems like we should expect memtrie to help with writes and not with reads. Reads were already optimised due to flat storage.

I probably don't understand this statement, because it seems incompatible with the discussion we had had with @tayfunelmas, [summarized above](https://github.com/near/nearcore/issues/11925#issuecomment-2284032050).

Storage reads were already a bit better with flat storage, but memtrie allows us to bypass any disk read, which is a lot of improvement even compared to the only-one-disk-read of flat storage. And on the other hand storage writes still require us to write to all of memtrie, flat storage and on-disk trie, which means that writes are actually a bit slower with memtrie than without.

This understanding is confirmed by the numbers I posted above, so I probably don't understand this sentence.

> Finally, there are some other concerns with reducing the gas costs. We should look at what the current compte costs are and instead consider reducing them instead.

The project that was given to me was to reduce gas costs. But as per the first message, I only looked at reducing compute costs for now, and all the "from" numbers given are for compute costs.

This being said, the numbers I suggested changing to are not only lower than our current compute costs, but also lower than our current gas costs. So after this lands and we confirm that it doesn't break anything (ie. after we're sure we won't need to revert the changes in a hurry), we will be able to consider also lowering the gas costs to the new value of the compute costs.

**Comment by Ekleog-NEAR** [comment]
> @Ekleog-NEAR when you run the experiments, did you include https://github.com/near/nearcore/commit/0c2374993fc74b57faf2bcdf5c7c73a37e82b75a ? According to @pugachAG this significantly improves memtrie performance.

I did not; IIRC it was still marked as draft when I ran the experiments. I'll run a new set of experiments and see how things evolve with it.

**Comment by Ekleog-NEAR** [comment]
Here are the results of the new set of experiments, run on fc87c851c635dd8fdbfc533dcb940125c9355859 so including the changes of 0c23749:

![chart (5)](https://github.com/user-attachments/assets/99abc7bb-cd2c-4a8a-80d1-24999bbd2bb2)
![chart (6)](https://github.com/user-attachments/assets/6b61422a-9617-44c9-80f8-d8e088906a92)
![chart (7)](https://github.com/user-attachments/assets/7fc2f3be-ed76-43d8-ac23-a813be91a1f0)

I'm actually not finding much difference between the results with 0c23749 and the results without it: the estimator results are almost always within one standard deviation of the previous results. My guess is that this is because the estimator does not actually exercise the prefetcher, and so the prefetcher is a no-op for it, while it originally brought slowdown to our real-world use case.

TL;DR: No changes with 0c23749, I still recommend the compute costs changes from https://github.com/near/nearcore/issues/11925#issuecomment-2299273181. The speedup from it is probably in parts that had never been measured by the estimator anyway, because it was too much contract-specific.

**Comment by Ekleog-NEAR** [comment]
Summary of the meeting with @akhi3030, @Longarithm and @pugachAG:
- We don't expect compute costs reductions for the write-specific costs, only for TTN which would then indirectly reduce write costs
- We should run an experiment: add metrics specifically for each cost we're looking to reduce as a histogram, and check that on a mainnet canary node the durations are as we would expect
- Yes TTN only ever touches memtrie, so we don't need to care about the 4k size split
- We currently don't charge TTN for reads because of flat storage, this is something we should fix now that we're significantly reducing compute costs

-----

Considering [this zulip thread](https://near.zulipchat.com/#narrow/stream/295302-general/topic/.E2.9C.94.20StorageHasKeyBase.2FByte), my assumption on `storage_has_key` was also correct.

So to sum up the next steps:
1. Split the costs that need to be different for values less than 4k into two costs
2. Reduce the costs that we can simply reduce
3. Add a new ReadTTN cost, with gas cost of 0 and compute cost of 6G, charged on TTN on reads (so that it doesn't charge more gas but only more compute costs)
4. Introduce ReadTTN to the parameter estimator
5. Add histogram metrics to all the costs we're considering changing
6. Run for a while on canary mainnet, check that the metrics are as expected
7. Assuming everything is fine, land
8. (Interestingly, if everything goes fine, we'll have compute costs lower than gas costs here. This being said, we should wait before lowering gas costs, especially because of things like ReadTTN)

With this, this issue has become a bit of a hydra, with lots of new heads popping out. I'd expect the remaining time needed to do it all to be around 3-4 weeks. If we skip steps 5&6, it'd reduce by ~1 week, and probably the same if we skip step 3&4 (and instead charge arbitrarily more on read costs). This being said, this will be my first time adding a new cost, I'll have to hope [the documentation we have](https://github.com/near/nearcore/blob/master/docs/architecture/gas/parameter_definition.md) is still up-to-date, even though it's more than a year old.

For now I'll proceed with doing all the steps, but feel free to let me know if you want me to skip some of them.

**Comment by Ekleog-NEAR** [comment]
Over the week-end I gave a bit more thought to the suggested cost changes, and I want to change the new TTN cost.

One thing I had initially forgotten was, that (Write)TTN cost must also include the cost for writing the updated node to disk, after updating the leaf value. So we need to plan for more variance in the numbers, because WriteTTN must include disk access time, and disk access has large variance. We don't need to do the same for ReadCachedTrieNode, because we need to write back the node only once anyway.

However, we don't need to handle the full latency that was handled before, because writes are batched at the end of the block processing. So the latency needs to be accounted for only once, and it already is in the Write*/Remove* costs.

With this in mind, here are my new suggested cost changes. I basically made the 6G of TTN into 20G, in order to account for writeback variance better:
- StorageHasKeyBase: 200,000,000,000 -> 8,000,000,000, without 4k split because checking whether the key exists should AFAIU never require falling back to disk (⚠️ assumption that needs to be confirmed by someone familiar with memtrie)
- StorageHasKeyByte: 30,790,845 -> 9,000,000, without 4k split (⚠️ same assumption)
- StorageReadBase: 200,000,000,000 -> same if the value is 4k or more; 9,000,000,000 if the value is strictly less than 4k
- StorageReadKeyByte: 30,952,533 -> same if the value is 4k or more; 10,000,000 otherwise
- StorageReadValueByte: 5,611,005 -> same if the value is 4k or more; 2,500,000 otherwise
- StorageWrite*: do not change
- StorageRemove*: do not change
- TouchingTrieNode: 110,000,000,000 -> 20,000,000,000, without 4k split (⚠️ same assumption)
- ReadCachedTrieNode: 2,280,000,000 -> 1,400,000,000, we probably do not need the 2x safety margin that we originally had, even though memtrie did not make things much faster the estimator is still around 1,070,000,000 only

**Comment by akhi3030** [comment]
This looks like a reasonable proposal to me.

**Comment by bowenwang1996** [comment]
If we plan to make changes to the `TrieAccountingCache`, should we avoid changing `ReadCachedTrieNode` for now? cc @akhi3030

**Comment by akhi3030** [comment]
good point, we will continue discussing and consider this.

**Comment by Ekleog-NEAR** [comment]
I just noticed an issue with reintroducing ReadTTN for memtrie reads. The details are [in this zulip thread](https://near.zulipchat.com/#narrow/stream/295302-general/topic/memtrie.20readTTN.2C.20rpc.20and.20archival.20nodes); it might result in a change in suggested costs changes.

**Comment by Ekleog-NEAR** [comment]
I also chatted with Akhi about `ReadCachedTrieNode` today. Basically, reducing `ReadCachedTrieNode` is not going to be more complex than not reducing it.

This being said, the gains are not that good. For a random tg receipt, the reduction would lead to a ~2% throughput gain. The gains would be similar for two kaiching receipts.

2% is neither a big nor a small number for the improvement of the throughput of the entire chain.

I can easily leave the cost reduction as-is (so, with the 2% throughput gains) until if/when we make changes to `TrieAccountingCache`, or remove it to make fewer changes (so, removing the 2% throughput gains).

Regardless of the choice, as we're not touching any gas costs, it'll be just as easy to change `TrieAccountingCache` later on, and we're not adding any tech debt as it's just a parameter change.

@bowenwang1996 What do you prefer?

**Comment by Ekleog-NEAR** [comment]
As part of the evaluation of #12044, I computed a higher bound on the benefits ReadTTN could give us. On the real-world use cases [with detailed data available here](https://docs.google.com/spreadsheets/d/1t6MPLVf4UYJr8vcw_VWRppUfZYvJHEUcKTYtziN0pLw/edit?usp=sharing), the benefits could not exceed 10-20%, as this is the remaining cost of the overestimated read costs.

I'd expect the benefits of just ReadTTN to be around 5-10% of the speed of the whole chain, because most likely, most nodes are at least at depth 12. This being said, the current 6G estimation for ReadTTN is probably also overestimated. If we also re-estimate ReadTTN, we might reach 10-15% of chain throughput improvements
